dashboard/app/models/stage.rb

```diff
@@ -75,6 +75,32 @@ def summarize
         levels: script.script_levels.to_a.select{|sl| sl.stage_id == id}.map(&:summarize),
     }
 
+    # Go through all levels.  If we find an assessment with a per_page value,
+    # then it's a long assessment: we assume it's the final level in the
+    # stage, and generate some placeholder levels for the subsequent pages
+    # of the long assessment.  Each of those levels will have the same basic
+    # URL as the first, but with ?page=1, ?page=2, etc.
+
+    levels = stage_data[:levels]
+
+    # We might append additional levels, so iterate over the original set.
+    levels[0..levels.length].each do |level|
+      if level[:kind] == "assessment"
```

Comment:
Hm, maybe the ScriptLevel#summarize method can generate the "additional
levels" also?

To avoid changing the api of ScriptLevel#summarize (or do all callers of
ScriptLevel#summarize want to see the extra levels)? you could add an
:extra_levels key which Stage#summarize could use.

I agree that it makes sense to keep this in a summarize method but maybe
more of it belongs to the Level/ScriptLevel summarize.


On Mon, Mar 14, 2016 at 4:57 AM, Brendan Reville <notifications@github.com>
wrote:

> In dashboard/app/models/stage.rb
> <https://github.com/code-dot-org/code-dot-org/pull/7257#discussion_r55989507>
> :
>
> > @@ -75,6 +75,32 @@ def summarize
> >          levels: script.script_levels.to_a.select{|sl| sl.stage_id == id}.map(&:summarize),
> >      }
> >
> > +    # Go through all levels.  If we find an assessment with a per_page value,
> > +    # then it's a long assessment: we assume it's the final level in the
> > +    # stage, and generate some placeholder levels for the subsequent pages
> > +    # of the long assessment.  Each of those levels will have the same basic
> > +    # URL as the first, but with ?page=1, ?page=2, etc.
> > +
> > +    levels = stage_data[:levels]
> > +
> > +    # We might append additional levels, so iterate over the original set.
> > +    levels[0..levels.length].each do |level|
> > +      if level[:kind] == "assessment"
>
> Got a thought on how to do this? I deliberately kept it away from the
> model because I didn't want any of the systems that rely on the actual
> models to know about these extra levels, and am deliberately generating
> them in the summarize family of functions so that they just magically
> appear closer to the progress-rendering.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/7257/files#r55989507>.
>